src/combinations.rs

```diff
@@ -53,7 +91,7 @@ impl<I> Iterator for Combinations<I>
     type Item = Vec<I::Item>;
     fn next(&mut self) -> Option<Self::Item> {
         if self.first {
-            if self.pool.is_done() {
+            if self.k() > self.n() {
```

Comment:
Can you walk me through why this line changed?

Reply:
Sure!

Previously, before my changes, the combined condition `first && pool.done` served to indicate insufficient items in the source iterator, to ever yield any elements from a `Combinations`. This worked because at construction of a `Combinations` we would ask for precisely `k` items and asking for the `k+1`th item would cause `pool.done` to be set.

With my changes we are reusing the `Combinations` (via the `reset()` method), and this reuse is 'triggered' by exhausting the source iterator. Therefore it's now possible for a `Combinations` to exist where all of the following hold:
  - source iterator is exhausted, i.e. `pool.done == true`
  - `k <= n == true`
  - we have not yielded any items yet, i.e. `first == true`

With the upshot being that `first && pool.done` no longer reliably indicates that we have no combinations to yield. 

Effectively `self.first && self.pool.done` was overloaded to mean `k < n` (i.e. 'no combinations are possible'), by relying on a particular way of constructing `Combinations`. Changing the second part of the condition to `k > n` means that we reliably detect an insufficient pool size once again.

Reply:
Additionally, for me, the (new) code above at lines 55 and 56 just reads more clearly as 'do a one time check that we have a big enough pool to yield combinations of length `k`'